src/cpu.c

```diff
@@ -440,33 +440,33 @@ static void aggregate(gauge_t *sum_by_state) /* {{{ */
  * current rate; each rate may be NAN. Calculates the percentage of each state
  * and dispatches the metric. */
 static void cpu_commit_one(int cpu_num, /* {{{ */
-                           gauge_t rates[static COLLECTD_CPU_STATE_MAX]) {
+                           gauge_t rates[static COLLECTD_CPU_STATE_MAX],
+                           gauge_t global_rate_sum) {
   metric_family_t fam = {
       .name = "system.cpu.utilization",
+      .help = "Difference in system.cpu.time since the last measurement, "
+              "divided by the elapsed time and number of logical CPUs",
+      .unit = "1",
       .type = METRIC_TYPE_GAUGE,
   };
 
   metric_t m = {0};
-  if (cpu_num == -1) {
-    metric_label_set(&m, "cpu", "total");
-  } else {
+  if (cpu_num != -1) {
     char cpu_num_str[32];
     ssnprintf(cpu_num_str, sizeof(cpu_num_str), "%d", cpu_num);
-    metric_label_set(&m, "cpu", cpu_num_str);
+    metric_label_set(&m, "system.cpu.logical_number", cpu_num_str);
   }
 
-  gauge_t sum = rates[COLLECTD_CPU_STATE_ACTIVE];
-  RATE_ADD(sum, rates[COLLECTD_CPU_STATE_IDLE]);
-
   if (!report_by_state) {
-    m.value.gauge = 100.0 * rates[COLLECTD_CPU_STATE_ACTIVE] / sum;
-    metric_label_set(&m, "state", cpu_state_names[COLLECTD_CPU_STATE_ACTIVE]);
-    metric_family_metric_append(&fam, m);
+    metric_family_append(
+        &fam, "system.cpu.state", cpu_state_names[COLLECTD_CPU_STATE_ACTIVE],
+        (value_t){.gauge = rates[COLLECTD_CPU_STATE_ACTIVE] / global_rate_sum},
+        &m);
   } else {
     for (size_t state = 0; state < COLLECTD_CPU_STATE_ACTIVE; state++) {
-      m.value.gauge = 100.0 * rates[state] / sum;
-      metric_label_set(&m, "state", cpu_state_names[state]);
-      metric_family_metric_append(&fam, m);
+      metric_family_append(&fam, "system.cpu.state", cpu_state_names[state],
+                           (value_t){.gauge = rates[state] / global_rate_sum},
```

Comment:
This is changing public collectd API, so both the name and range changes require documentation updates.

Reply:
I feel that the description (line 447) sufficiently explains the metric. Where else would you like to document it?

Yes, this is a change in the metric, just like any other name change also is. I'm trying to get this into shape before we do a release, so we don't have to change this later. My hope is that we get the core plugins (cpu, memory, disk, df) into a shape where we don't have to make any changes to the metrics, even after a "release candidate".

Reply:
You have a point, we should start writing a "migration guide" that summarizes the changes between collectd 5 and 6, and gives guidance on how to upgrade an existing setup.

re no demand: I know of two companies that are phasing out their use of collectd in favor of an OpenTelemetry based system. One company took a detour over Prometheus, the other didn't.

re too much change: putting the plugin updates in with the new `unit` field is not ideal. I'm happy to make separate PRs if you prefer.

re backwards compatibility: collectd 6 is **not** meant to be a drop-in replacement for collectd 5. It is a new major version specifically so that we are free to change data structures, behavior, integration points, etc. as needed to get to a base that we can build off of for the foreseeable future.

I think it is key to also provide interoperability between collectd 5 and 6. At the very least, a collectd 6 server should be able to accept metrics using the *network plugin*, so that not the entire fleet has to be upgraded in lockstep.

Reply:
Prometheus does not accept metrics or labels with dots in their names, so this breaks `write_prometheus`.  At least for now, I would think working (well) with Prometheus to be more important than OpenTelemetry spec support. :-/

> re no demand: I know of two companies that are phasing out their use of collectd in favor of an OpenTelemetry based system.

Do you know whether their reasons were related to OpenTelemetry spec support, or those systems (which ones?) just being more actively developed & having releases in past few years?

Reply:
Ah, Google.  Then it's probably also "metrics collector has open network sockets and managed languages like Golang are better for security" + "let's avoid GPL by rewriting whole thing as we have resources for it".

(Downside of those Golang projects is that they do not allow linking to a non-Golang library, so *everything* required by a plugin really needs to be written from scratch in that language.   Which is not really a feasible option for something like Sysman functionality.)

Reply:
> You have a point, we should start writing a "migration guide" that summarizes the changes between collectd 5 and 6, and gives guidance on how to upgrade an existing setup.

Yes, but I guess I cannot ask that in this PR. :-)

> re too much change: putting the plugin updates in with the new unit field is not ideal. I'm happy to make separate PRs if you prefer.

Having one plugin in that PR as example of the usage for the new APIs, would be fine (e.g. memory plugin as changes in that seem smallest), with rest of plugins coming in other PR(s).

Reply:
> Having one plugin in that PR as example of the usage for the new APIs, would be fine (e.g. memory plugin as changes in that seem smallest), with rest of plugins coming in other PR(s).

Done, plugin updates moved to: #4211